### PR TITLE
feat(git): `tools git branch-gc` — clean up stale & merged local branches (squash-aware)

### PR DESCRIPTION
**Pull request by genesiscz**
## What

Adds `tools git branch-gc` — a squash-aware local-branch hygiene **subcommand of the `git` tool**. It lists every local git branch, classifies each by how safe it is to delete, shows last-commit age + ahead/behind, and lets you delete the safe ones interactively (clack multiselect) or non-interactively (`--yes`). Default behavior is a read-only listing.

## Why

`git branch --merged` only catches branches joined by a real merge commit. The dominant workflow today is **squash-merge** (GitHub "Squash and merge", `git merge --squash`), which leaves the local branch looking "not fully merged" even though every change is already in `master` — so `git branch -d` refuses, and people either keep dozens of dead branches or reach for `-D` and risk nuking real work.

## Classification

Each branch gets exactly one status, in priority order:

- `current` / `base` — never deletable.
- `merged` — tip is an ancestor of base (`git merge-base --is-ancestor`); catches real merges + fast-forwards.
- `squash-merged` — the load-bearing heuristic. Synthesizes a single squashed commit from the branch's tree on top of its merge-base (`git commit-tree`, with a fixed identity so it's deterministic on any machine/CI), then asks `git cherry base squashed` whether base already contains the combined patch (leading `-`). This is what plain `git cherry base branch` gets wrong on squashes.
- `gone` — upstream tracking branch was deleted on the remote (`%(upstream:track)` = `[gone]`).
- `stale` — no commit newer than `--stale-days` (default 90). Offered interactively, never auto-deleted.
- `active` — none of the above. Never offered.

`--yes` deletes the safe set (`merged` + `squash-merged` + `gone`). The current and base branches are NEVER deletable.

## CLI

```
tools git branch-gc                       # list & classify (read-only default)
tools git branch-gc --yes                 # delete merged + squash-merged + gone, no prompts
tools git branch-gc --base main --stale-days 30
tools git branch-gc --json                # machine-readable, no deletion
tools git branch-gc --no-dry-run          # opt into interactive multiselect in a TTY
tools git branch-gc -C <path>             # run against the repo at <path>
```

## Layout & design

- Lives under the existing `git` tool: the controller is `src/git/commands/branch-gc.ts` (`registerBranchGcCommand`), and the pure classification logic is in `src/git/lib/branch-gc/classify.ts`. Registered in `src/git/index.ts` alongside `commits` / `configure-authors` / `configure-workitem-patterns`.
- All git classification logic is pure, exported functions; the current time is injected (`nowEpoch`) so staleness is deterministic and unit-testable.
- **Git plumbing consolidated onto the shared `@app/utils/git`** (`createGit().executor`) — no tool-local `runGit`. The squash-detection `commit-tree` writes an unreachable (gc-pruned) object with a fixed identity, so the result is portable across machines/CI rather than depending on ambient git config.
- Result channel is clean: the table / `--json` go to stdout via `out.println` / `out.result`; all framing, summaries, prompts, and diagnostics go to stderr.

## Tests

`bun test src/git/lib/branch-gc` — 12 hermetic tests against throwaway tmp git repos (no network/clipboard/model). Covers merged, squash-merged (with an assertion that plain `git cherry` would miss it), active, the stale↔active boundary flip via injected `now`, current/base exclusion, ahead/behind, gone-upstream, base auto-detection (master/main fallback, no-base, detached HEAD).

The repo-building git spawns pass an explicit `env` (`GIT_CONFIG_GLOBAL=/dev/null`, fixed identity + commit dates) — Bun.spawn snapshots `process.env` at process start and ignores later mutations, so per-call env is required for hermetic, deterministic commits.

## Verification

- `bunx biome check src/git` → exit 0
- `bun test src/git/lib/branch-gc` → 12 pass (incl. squash-merge case under a stripped, no-identity env)
- `./tools git branch-gc --help` → exit 0
- `./tools git branch-gc --json` against a live repo → valid JSON classification, exit 0

## Summary by CodeRabbit

* **New Features**
  * Added `git branch-gc` command to identify and remove stale or merged local Git branches.
  * Classifies branches into merged, squash-merged, gone, stale, active (including current/base), with optional age and ahead/behind counts.
  * Supports JSON output, non-interactive `--yes` deletion, interactive selection, and dry-run preview.
  * Base branch auto-detection (master/main) or explicit override; deletes only “safe” branches.
* **Tests**
  * Added a Bun test suite covering branch classification and base/current-branch detection.

### CODE REVIEW

**Comment by genesiscz** [comment]
@coderabbitai review

File: src/git/lib/branch-gc/classify.ts [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+
+    const mb = mbRes.stdout.trim();
+    if (!mb) {
+        return false;
+    }
+
+    const countRes = await runGit(cwd, ["rev-list", "--count", `${mb}..${branch}`]);
+    if (Number.parseInt(countRes.stdout.trim(), 10) === 0) {
+        return false;
+    }
```
[the remark this answers is not in the document]
Comment (genesiscz) [reply src/git/lib/branch-gc/classify.ts]: /gemini Fixed in [5a9229f](https://github.com/genesiscz/GenesisTools/commit/5a9229f2306848cae1b87880c55968def02d3206) - added a countRes.code !== 0 guard to the rev-list short-circuit so a non-zero exit returns false instead of falling through into commit-tree/cherry on bogus state (parseInt of empty is NaN, and NaN === 0 is false). Also matches every other runGit check in this file.

File: src/branch-gc/lib/classify.ts [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+        if (await isSquashMerged(o.cwd, o.base, name)) {
+            return "squash-merged";
+        }
+
+        if (goneSet.has(name)) {
+            return "gone";
+        }
+
+        if (age > staleDays) {
+            return "stale";
```
[the remark this answers is not in the document]
Comment (genesiscz) [reply src/branch-gc/lib/classify.ts]: @coderabbitai Fixed in [5a9229f](https://github.com/genesiscz/GenesisTools/commit/5a9229f2306848cae1b87880c55968def02d3206) - changed age > staleDays to age >= staleDays. The help text says branches with no commit newer than this many days are stale, so a branch whose newest commit is exactly staleDays old should be stale. ageDays is floored so the boundary is reachable, and the existing boundary test only probes ageDays-1 / ageDays+1 so it stays green.

File: src/branch-gc/lib/classify.test.ts [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+const SECONDS_PER_DAY = 86_400;
+// Fixed wall-clock anchor so every committed date is deterministic.
+const ANCHOR_EPOCH = 1_700_000_000;
+
+// Hermetic git env: no global/system config (gpgsign, templateDir, hooks),
+// fixed identity. Bun.spawn snapshots env at process start and ignores later
+// process.env mutations, so we MUST pass this explicitly on every spawn — the
+// repo-building git() helper below does exactly that.
+const BASE_GIT_ENV: Record<string, string> = {
+    ...(process.env as Record<string, string>),
```
[the remark this answers is not in the document]
Comment (genesiscz) [reply src/branch-gc/lib/classify.test.ts]: @coderabbitai Fixed in [5a9229f](https://github.com/genesiscz/GenesisTools/commit/5a9229f2306848cae1b87880c55968def02d3206) - dropped the unsafe process.env cast and the explicit annotation, letting TS infer. Bun.spawn's env accepts Record<string, string | undefined>, so the spread type-checks without the cast, and this aligns with the repo's no-inline-type-casts rule.

File: src/branch-gc/index.ts [hunk 4]
```diff
@@ -0,0 +1,214 @@
+import { logger, out } from "@app/logger";
```
[the remark this answers is not in the document]
Comment (genesiscz) [reply src/branch-gc/index.ts]: @coderabbitai Fixed in [5a9229f](https://github.com/genesiscz/GenesisTools/commit/5a9229f2306848cae1b87880c55968def02d3206) - the top-level catch now logs the raw error via the scoped logger before printing a concise stderr message: log.error({ error: err }, 'branch-gc failed') then out.error(...). out.error is the user-facing stderr channel only, so the file log was never seeing the error. Declining the shebang note - tools here are launched via bun run and several entrypoints intentionally omit it, so the file start stays unchanged.

File: src/branch-gc/index.ts [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+    }
+
+    const deleted = await deleteMany(cwd, choice);
+    out.log.success(`Deleted ${deleted}/${choice.length} branch(es).`);
+}
+
+main().catch((err) => {
+    out.error(String(err));
+    process.exit(1);
+});
```
[the remark this answers is not in the document]
Comment (genesiscz) [reply src/branch-gc/index.ts]: @coderabbitai Fixed in [5a9229f](https://github.com/genesiscz/GenesisTools/commit/5a9229f2306848cae1b87880c55968def02d3206) - same fix as the line-1 thread. The catch now logs the raw error object first via the scoped logger, then a concise stderr message: log.error({ error: err }, 'branch-gc failed'); out.error(err instanceof Error ? err.message : String(err)); process.exit(1). Matches the prefer error: err and triage-from-logs-alone guidelines.

File: src/branch-gc/lib/classify.ts [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+    }
+
+    const tree = treeRes.stdout.trim();
+    const squashRes = await runGit(cwd, ["commit-tree", tree, "-p", mb, "-m", "_"]);
+    if (squashRes.code !== 0) {
+        return false;
+    }
+
+    const squashed = squashRes.stdout.trim();
+    const cherryRes = await runGit(cwd, ["cherry", base, squashed]);
```
[the remark this answers is not in the document]
Comment (genesiscz) [reply src/branch-gc/lib/classify.ts]: @coderabbitai Fixed in [5a9229f](https://github.com/genesiscz/GenesisTools/commit/5a9229f2306848cae1b87880c55968def02d3206) - isSquashMerged now sandboxes the commit-tree/cherry object writes. runGit gained an optional env arg; the synthesized commit goes into a throwaway GIT_OBJECT_DIRECTORY with the real repo as a read-only GIT_ALTERNATE_OBJECT_DIRECTORIES (resolved via rev-parse --path-format=absolute --git-path objects so worktrees work), then the temp dir is removed. Verified: count-objects is unchanged before/after and squash detection still returns true.

File: src/git/lib/branch-gc/classify.test.ts [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+                staleDays: 90,
+            });
+            const feat = infos.find((b) => b.name === "feat/diverged");
+            expect(feat?.ahead).toBe(2);
+            expect(feat?.behind).toBe(1);
+        } finally {
+            rmSync(dir, { recursive: true, force: true });
+        }
+    });
+});
```
[the remark this answers is not in the document]
Comment (genesiscz) [reply src/git/lib/branch-gc/classify.test.ts:254]: @coderabbitai Fixed in [5a9229f](https://github.com/genesiscz/GenesisTools/commit/5a9229f2306848cae1b87880c55968def02d3206) - added a gone-status test that creates a real bare remote, pushes a tracked branch, deletes the upstream, fetches --prune, and asserts status gone + deletable true. Note the attached CI annotation (line 111 expected squash-merged, received active) was stale - the suite is green: now 12 pass / 0 fail.

File: src/git/lib/branch-gc/classify.test.ts [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+
+    it("throws BaseNotFoundError for an explicit base that does not exist", async () => {
+        const dir = await makeRepo();
+        try {
+            await expect(detectBase(dir, "nope")).rejects.toThrow(/does not exist/);
+        } finally {
+            rmSync(dir, { recursive: true, force: true });
+        }
+    });
+});
```
[the remark this answers is not in the document]
Comment (genesiscz) [reply src/git/lib/branch-gc/classify.test.ts:309]: @coderabbitai Fixed in [5a9229f](https://github.com/genesiscz/GenesisTools/commit/5a9229f2306848cae1b87880c55968def02d3206) - added three edge-case tests: auto-detect main when there is no master, BaseNotFoundError when neither master nor main exists, and getCurrentBranch returning '' on a detached HEAD. They follow the existing helper/cleanup style and all pass.

File: src/branch-gc/lib/classify.ts [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+        return false;
+    }
+
+    const treeRes = await runGit(cwd, ["rev-parse", `${branch}^{tree}`]);
+    if (treeRes.code !== 0) {
+        return false;
+    }
+
+    const tree = treeRes.stdout.trim();
+    const squashRes = await runGit(cwd, ["commit-tree", tree, "-p", mb, "-m", "_"]);
```
[the remark this answers is not in the document]
Comment (genesiscz) [reply src/branch-gc/lib/classify.ts]: /gemini Won't fix - git commit-tree does NOT require a configured identity (unlike git commit). Verified empirically with the identity env vars unset and HOME plus global/system config pointed at /dev/null: commit-tree still returns a SHA with exit 0 because it synthesizes a default user-at-hostname ident. Only git commit (with user.useConfigOnly) hard-fails on a missing identity. So isSquashMerged will not silently fail on fresh CI runners; the suggested -c overrides fix a non-bug.

File: src/git/lib/branch-gc/classify.ts [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+/**
+ * Classify every local branch into exactly one status, applying the spec's
+ * priority order. Pure w.r.t. time — uses `opts.nowEpoch`, never `Date.now()`.
+ * Sorted: safe-to-delete first, then stale/active by age desc, base & current last.
+ */
+export async function classifyBranches(opts: ClassifyOptions): Promise<BranchInfo[]> {
+    const [branches, gone] = await Promise.all([listLocalBranches(opts.cwd), getUpstreamGone(opts.cwd)]);
+
+    const infos = await Promise.all(branches.map((b) => classifyOne(b, opts, gone)));
+
```
[the remark this answers is not in the document]
Comment (genesiscz) [reply src/git/lib/branch-gc/classify.ts:110]: /gemini By design - keeping the parallel classification. A fully sequential per-branch loop would throw away the win for typical repos. The genuine concern is unbounded fan-out (each branch spawns up to ~5 git subprocesses, so a 100-branch repo can hit ~500 concurrent spawns); if that ever bites we'd bound it with the repo's existing concurrentMap helper in src/utils/async.ts (default concurrency 5), not drop to fully sequential. Declining the full-sequential change.

File: src/branch-gc/index.ts [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+
+async function deleteBranch(cwd: string, name: string): Promise<boolean> {
+    const shaRes = await runGit(cwd, ["rev-parse", "--short", name]);
+    const sha = shaRes.stdout.trim();
+
+    const res = await runGit(cwd, ["branch", "-D", name]);
+    if (res.code === 0) {
+        log.out.success(`Deleted ${pc.bold(name)} ${pc.dim(`(restore: git branch ${name} ${sha})`)}`);
+        return true;
+    }
```
[the remark this answers is not in the document]
Comment (genesiscz) [reply src/branch-gc/index.ts:86]: /gemini By design - the malformed restore hint is unreachable, so the stated bug can't occur. If the branch was concurrently deleted, rev-parse --short fails (exit 128, empty stdout) AND branch -D fails in lockstep (exit 1, branch not found). The restore hint is printed only inside the res.code === 0 success path, so a deleted branch takes the error path and an empty sha can never reach the hint string. The conditional-sha form is fine as clarity but isn't fixing a real output bug.

File: src/branch-gc/index.ts [hunk 12]
```diff
@@ -0,0 +1,214 @@
+import { logger, out } from "@app/logger";
+import { isInteractive, runTool, suggestCommand } from "@app/utils/cli";
+import * as p from "@clack/prompts";
+import { Command } from "commander";
+import pc from "picocolors";
```
[the remark this answers is not in the document]
Comment (genesiscz) [reply src/branch-gc/index.ts]: @coderabbitai Won't fix - the premise that chalk is the house style is inverted. In this repo picocolors has 173 importers vs chalk's 65; both coexist as first-class and picocolors is the lighter/faster choice. There's no style drift to correct, and per our surgical-changes / match-existing-style rule, swapping a working color lib is exactly the kind of unrequested churn to avoid.

File: src/git/lib/branch-gc/classify.ts [hunk 13]
```diff
@@ -0,0 +1,260 @@
...
+        throw new BaseNotFoundError(`Base branch "${explicit}" does not exist locally.`);
+    }
+
+    for (const candidate of ["master", "main"]) {
+        if (await localBranchExists(cwd, candidate)) {
+            return candidate;
+        }
+    }
+
+    throw new BaseNotFoundError("Could not auto-detect a base branch (no local master/main). Pass --base <branch>.");
```
[the remark this answers is not in the document]
Comment (genesiscz) [reply src/git/lib/branch-gc/classify.ts]: @coderabbitai By design - already guarded upstream. Exit codes are correct (missing ref 1, non-repo 128), but detectBase/localBranchExists are only reached after the entrypoint guard at src/branch-gc/index.ts:116: if not inside a work tree (rev-parse --is-inside-work-tree), it errors and exits. So by the time detectBase runs we're guaranteed inside a repo and the exit-128 not-a-repository path can't occur. The only other caller is the test suite (controlled repos). Collapsing non-zero to false is intentional given that precondition; distinguishing 1 vs 128 would only matter if these libs were reused outside this entrypoint (YAGNI today).

**Comment by genesiscz** [comment]
@coderabbitai review

**Comment by genesiscz** [comment]
@coderabbitai review

File: src/git/commands/branch-gc.ts [hunk 14]
```diff
@@ -0,0 +1,229 @@
...
+    detectBase,
+    getCurrentBranch,
+} from "@app/git/lib/branch-gc/classify";
+import { logger, out } from "@app/logger";
+import { isInteractive, suggestCommand } from "@app/utils/cli";
+import { createGit } from "@app/utils/git";
+import type { Storage } from "@app/utils/storage";
+import * as p from "@clack/prompts";
+import type { Command } from "commander";
+import pc from "picocolors";
```
[the remark this answers is not in the document]
Comment (genesiscz) [reply src/git/commands/branch-gc.ts:22]: @coderabbitai By design - the premise that chalk is the repo standard is inverted. In this repo picocolors has 186 importers vs chalk's 66; both coexist as first-class and picocolors is the lighter/faster pick. The PR #216 learning does not make chalk mandatory, and per the surgical-changes / match-existing-style rule, swapping a working color lib is exactly the unrequested churn to avoid. Keeping picocolors.

File: src/git/commands/branch-gc.ts [hunk 15]
```diff
@@ -0,0 +1,229 @@
...
+
+async function deleteBranch(cwd: string, name: string): Promise<boolean> {
+    const { executor } = createGit({ cwd });
+
+    const shaRes = await executor.exec(["rev-parse", "--short", name]);
+    const sha = shaRes.stdout;
+
+    const res = await executor.exec(["branch", "-D", name]);
+    if (res.success) {
+        log.out.log.success(`Deleted ${pc.bold(name)} ${pc.dim(`(restore: git branch ${name} ${sha})`)}`);
```
[the remark this answers is not in the document]
Comment (genesiscz) [reply src/git/commands/branch-gc.ts]: @coderabbitai Fixed in [c921cce](https://github.com/genesiscz/GenesisTools/commit/c921cce5101d27caffd7e2cdf5db5bbe2a846af1) - changed to out.log.success to match the file's own convention. One correction on the premise: there is no runtime-crash risk. log is logger.scoped('branch-gc').log, typed pino.Logger plus { out: Out; tee: Out }, so log.out is always a defined Out and log.out.log.success type-checks and resolves (tsgo is clean). The real issue is consistency - the verbose mirrored form was the only such call site in src/, so I switched it to the plain destructured out used everywhere else in the file.

File: src/git/commands/branch-gc.ts [hunk 16]
```diff
@@ -0,0 +1,229 @@
...
+        return true;
+    }
+
+    log.out.error(`Failed to delete ${name}: ${res.stderr || "unknown git error"}`);
```
[the remark this answers is not in the document]
Comment (genesiscz) [reply src/git/commands/branch-gc.ts]: @coderabbitai Fixed in [c921cce](https://github.com/genesiscz/GenesisTools/commit/c921cce5101d27caffd7e2cdf5db5bbe2a846af1) - changed to out.error to match lines 109/116/126/225. Same note as the sibling thread: log.out is always defined (scoped log is typed with { out: Out }), so there was no undefined-access crash - log.out.error type-checked and resolved fine. The fix is purely for internal consistency, since this was one of only two log.out.* call sites in the whole src/ tree.
